File: src/main/scala/vggames/task/IndexedTask.java
package vggames.task;

import vggames.task.status.Error;

public final class IndexedTask implements TaskWithDescription {

	private final TaskWithDescription delegate;
	private final int index;

	public IndexedTask(final TaskWithDescription delegate, final int index) {
		this.delegate = delegate;
		this.index = index;
	}

	public int getIndex() {
		return index;
	}

	public JudgedTask judge(final String challenge) {
		try {
			return delegate.judge(challenge);
		} catch (Exception e) {
			return new Error(e);
		}
	}

	public String getDescription() {
		return delegate.getDescription();
	}

	public String getChallenge() {
		return delegate.getChallenge();
	}

	@Override
	public String toString() {
		return delegate.toString();
	}

	public String getGroupName() {
		return delegate.getGroupName();
	}
}

File: src/main/scala/vggames/shared/task/IndexedTask.scala
package vggames.shared.task;

import vggames.shared.task.status.Error

class IndexedTask[T](delegate : TaskWithDescription[T], index : Int) extends TaskWithDescription[T] {

  def getIndex : Int = index

  def judge(challenge : String) : JudgedTask = {
    try {
      delegate.judge(challenge)
    } catch { case e : Throwable => new Error(e) }
  }

  def resource = delegate.resource

  def getDescription : String = delegate.getDescription

  def getChallenge : String = delegate.getChallenge

  override def toString : String = delegate.toString

  def getGroupName : String = delegate.getGroupName

  def getGroupCode = delegate.getGroupCode

  override def extraData = delegate.extraData

}
